src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreCycleDetector.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.PackageRestore
+{
+    internal sealed class NuGetRestoreCycleDetector
+    {
+        private readonly object _lock;
+        private readonly Queue<byte[]> _values;
+
+        /// <summary>
+        ///     Fixed size of the numbers of values to store.
+        /// </summary>
+        /// <remarks>
+        ///     This represents how deep to search for hash1 cycle.
+        /// </remarks>
+        public int Size { get; private set; }
+
+        public NuGetRestoreCycleDetector(int size)
+        {
+            _lock = new object();
+            _values = new Queue<byte[]>();
+            Size = size;
+        }
+
+        /// <summary>
+        ///     Validate if hash1 cycle exist. 
+        ///     If no cycle exist, then the hashValue is stored.
+        /// </summary>
+        /// <param name="hashValue">This is hashValue that is used to verify if it exists in previous restores</param>
+        /// <returns>True if it contains hash1 cycle, otherwise false</returns>
+        public bool ComputeCycleDetection(byte[] hashValue)
+        {
+            lock (_lock)
+            {
+                foreach (var b in _values)
+                {
+                    if (AreEqual(b, hashValue))
+                    {
+                        return true;
+                    }
+                }
```

Comment:
This loop detects whether the current hash has been seen within the last N hashes.

You could trigger this by adding a package reference, then removing it. Once removed, the project is back in the prior state and restore should complete.

With this change, I believe we would not restore after removing the recently added package reference.

As an alternative, consider something like:

- store the last N hashes
- each time a new hash is produced
  - append it to the list
  - discard any hashes added longer than some threshold ago
  - walk back through the list, looking for cycles of repeating patterns of length 2, 3, 4, ...
  - if a repeating pattern is found, there is a cycle

I'm not 100% sure about that approach, but I think it would be more robust than the approach here.

Reply:
> I am wondering if NuGet can easily detect a bad package

What would be a bad package in this case?

If you are referring to a package with props/targets that change the inputs, the MSBuild side is hidden from NuGet, so NuGet has no means of figuring it out, so this logic is technically handling that.

Reply:
> I modified the logic a little to detect the cycle only when counter == Size.

I'm concerned that this is only matching values and not looking for patterns.

For example (where each letter represents a hash):

- A B A B A B A B A B A B
- A B C A B C A B C A B C
- A B A C A D E A F A G A

The first two sequences look more like infinite cycles to me than the third one, yet the third would trip the check here.

Reply:
@nkolev92 do you think these patterns are possible to appear ?

Reply:
In theory, you can come up with packages/authoring that has a pattern of any length. 

In my experience, I've only seen  A B A B patterns.  

The most common occurrences of continuous restores is the dynamically generate version one. So maybe we try to detect situations where say a bunch of consecutive design time builds lead to a restore nomination. If we want to special case that, we can probably use the NuGet start/stop restore markers for help.

Reply:
@nkolev92  if we are in a loop, does it mean that we already completed NuGet restore ?

Reply:
yep. 

The loop is effectively, design time build which nominates, restore, design time build which nominates, restore.

Reply:
@nkolev92 at the project system level we don't know what package is causing the loop. We can show a yellow bar to notify the user that one of the nuget packages is invalid. Is there a way to identify the package from the NuGet side ? so we can get the name of the package from NuGet and use it in the yellow bar ?

Reply:
> @nkolev92 at the project system level we don't know what package is causing the loop. We can show a yellow bar to notify the user that one of the nuget packages is invalid and is causing a nuget restore cycle.

Yep, that's in line with what I'd expect based on how MSBuild works :) 

>  Is there a way to identify the package from the NuGet side ? so we can get the name of the package from NuGet and use it in the yellow bar ?

NuGet doesn't really anything MSBuild in VS, so there's pretty much no way of figuring it out.

Reply:
Done.
An InfoBar will be displayed in the main window with the message "NuGet restore detected a cycle. NuGet Error NU1108"